Port the transactions queue from go-ethereum to status-go

adambabik opened the pull request:
As we introduced upstream RPC server in #193, we also need to make sure that sending a transaction using any transport goes through the queue and complete/discard flow.

### Changes
1. As I changed `geth/node/txqueue.go` to `geth/node/txqueue_manager.go` and copied `txqueue.go` from go-ethereum, it might introduce some confusion when looking at the changes.
2. Changes to go-ethereum will be submitted in another PR in our go-ethereum fork.

### Todo:
* [x] decide which transport use in `TxQueueManager.CompleteTransaction`
* [ ] prepare a diagram of queuing process and when it interacts with go-ethereum
* [x] fix tests
* [x] code clean up

adambabik commented on geth/jail/jail.go:
```diff
@@ -201,18 +201,12 @@ func (jail *Jail) Send(call otto.FunctionCall) (response otto.Value) {
-		default:
-			res, resErr = jail.policy.ExecuteOtherTransaction(req, call)
-		}
+		// TODO(adam): these calls should be called async and results collected
```
@tiabc am I right here?

adambabik replied:
I see. Hm, I think it's safer to leave it for the time being as we have started testing the upstream RPC client and new commands will require filtering.

tiabc replied:
I'd remove this TODO and leave just a note about possible optimisation here and explaining why it's not currently done this way.

adambabik commented on geth/jail/jail.go:
```diff
@@ -239,99 +233,6 @@ func (jail *Jail) Send(call otto.FunctionCall) (response otto.Value) {
-}
-
-// preProcessRequest pre-processes a given RPC call to a given Otto VM
-func preProcessRequest(vm *otto.Otto, req common.RPCCall) (string, error) {
```
Moved to `execution_policy.go`.

adambabik commented on geth/jail/jail.go:
```diff
@@ -239,99 +233,6 @@ func (jail *Jail) Send(call otto.FunctionCall) (response otto.Value) {
-}
-
-// postProcessRequest post-processes a given RPC call to a given Otto VM
-func postProcessRequest(vm *otto.Otto, req common.RPCCall, messageID string) {
```
Moved to `execution_policy.go`.

adambabik commented on geth/jail/jail.go:
```diff
@@ -239,99 +233,6 @@ func (jail *Jail) Send(call otto.FunctionCall) (response otto.Value) {
-	}
-}
-
-func sendTxArgsFromRPCCall(req common.RPCCall) status.SendTxArgs {
```
Moved to `execution_policy.go`.

adambabik commented on geth/node/txqueue.go:
```diff
@@ -1,194 +1,254 @@
 package node
```
Diff for this file is messy because I moved the original `txqueue.go` to `txqueue_manager.go` and created `txqueue.go` as a new file.

influx6 commented on cmd/statusd/utils.go:
```diff
@@ -972,7 +959,7 @@ func testCompleteMultipleQueuedTransactions(t *testing.T) bool {
 
 		time.Sleep(1 * time.Second) // make sure that tx complete signal propagates
```
Do we have confidence that this will always work, do we need to find a means to guarantee that signal has propagated like we need to do for node sync?

influx6 replied:
With the sync PR  we would not need to worry about this anymore.

influx6 commented on geth/node/txqueue_manager.go:
```diff
@@ -0,0 +1,360 @@
+	ctx2, cancel2 := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel2()
+
+	var result json.RawMessage
```
What happens to result here, seems like it's used to receive response but not saved, or returned anywhere?

adambabik replied:
Actually, we don't need to compare it as it's not possible it will change. Following https://github.com/ethereum/go-ethereum/blob/bc0e6a5e68904b2a1251240b85dd97dc8fa07e11/ethclient/ethclient.go#L426-L432

adambabik commented on cmd/statusd/utils.go:
```diff
@@ -972,7 +959,7 @@ func testCompleteMultipleQueuedTransactions(t *testing.T) bool {
 
 		time.Sleep(1 * time.Second) // make sure that tx complete signal propagates
```
I don't think so. If you have figured out a better way to ensure that the node is in sync, this line should be fixed as well. But I think it's not in scope of this PR.

adambabik commented on geth/node/txqueue_manager.go:
```diff
@@ -0,0 +1,360 @@
+	ctx2, cancel2 := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel2()
+
+	var result json.RawMessage
```
Good catch. As `eth_sendRawTransaction` returns a transaction hash, I need to read it and validate that both hashes match, i.e. `result["result"] == signedTx.Hash()`.

adambabik commented on cmd/statusd/library.go:
```diff
@@ -212,7 +212,7 @@ func Logout() *C.char {
 //export CompleteTransaction
 func CompleteTransaction(id, password *C.char) *C.char {
-	txHash, err := statusAPI.CompleteTransaction(C.GoString(id), C.GoString(password))
+	txHash, err := statusAPI.CompleteTransaction(common.QueuedTxID(C.GoString(id)), C.GoString(password))
```
It introduces quite a lot of boilerplate and currently brings no value. We can think about it but I don't have any arguments for keeping it.

adambabik commented on geth/api/api.go:
```diff
@@ -34,6 +36,11 @@ func (api *StatusAPI) JailManager() common.JailManager {
+// TxQueueManager returns reference to account manager
+func (api *StatusAPI) TxQueueManager() common.TxQueueManager {
+	return api.b.TxQueueManager()
+}
```
You're right. It's used only to access `TxQueue` actually but I guess it also should not be exposed.

adambabik commented on geth/api/backend_txqueue_test.go:
```diff
@@ -53,22 +51,25 @@ func (s *BackendTestSuite) TestSendContractTx() {
 			log.Info("trying to complete with no user logged in")
-			txHash, err = s.backend.CompleteTransaction(event["id"].(string), TestConfig.Account1.Password)
+			txHash, err = s.backend.CompleteTransaction(
+				common.QueuedTxID(event["id"].(string)), TestConfig.Account1.Password)
```
Overall I agree, it's easier to read and edit. However, in my opinion, it's fine if the last argument is a message like in this example:
```go
t.EqualError(err, txError.Error(), 
  fmt.Sprintf("some long message with formating: %v", tx))
```

After coding in Python for a few years, I might be biased though :P

tiabc replied:
I wouldn't say so. When you introduce a new nested indenting level without a closing structure on the initial indenting level, you can't quickly locate the end of the function call, your eyes are looking below and below searching for the closing parenthesis. And in case of having each argument on a separate line you can quickly figure out where your function call ends. You also don't guess if it's an indenting error or by design.

By the way, Go doesn't enforce line width and keeping long lines is sometimes more readable than splitting them.

adambabik replied:
I think it's difficult to say for sure that it's the way everyone's eyes work. As I said, this type of code in some other languages and environments, like Python, is something normal. 

Consistency and project standards are more important and if we decided to put each param in a line line in such a case, then we should enforce it everywhere and I am willing to do so.

After we make the tests passing in CI, we can curate a list of linters and add `make lint` to the CI flow. It should handle such stylistic issues.

tiabc replied:
It's not the rule yet but I'd like to get it a try and cancel it if most people say they don't like it.

adambabik commented on geth/api/backend_txqueue_test.go:
```diff
@@ -344,25 +346,25 @@ func (s *BackendTestSuite) TestCompleteMultipleQueuedTransactions() {
 		if envelope.Type == node.EventTransactionQueued {
 			event := envelope.Event.(map[string]interface{})
-			txID = event["id"].(string)
+			txID := common.QueuedTxID(event["id"].(string))
```
Yeah, I don't see any argument for having this either. I ported it from go-ethereum implementation but it looks like premature optimization/security.

adambabik commented:
@tiabc sorry, I forgot to reply. 

> If you're introducing mocks, you should specify how to perform common operations with them. The best option is to make those commands in Makefile.

It's done, check out `make mock` and `make mock-install`.

> Could you also check (and better write a test) how eth_syncing behaves if UpstreamConfig.Enabled. status-react relies on it in a sense of showing blockchain synchronisation progress. Ideally, it shouldn't be broken.

Yeah, it's working fine, I checked manually. I don't see how these changes could affect `eth_syncing`. Also, I am not sure we should write such a general test here.

> Please, run make lint and fix errors.

I fixed what I found relevant to these changes but this command prints insane amount of alerts, even in internal go files. We should revise its purpose and just select the most important rules.

tiabc commented:
@adambabik 

> It's done, check out make mock and make mock-install.
Thanks.

> Yeah, it's working fine, I checked manually. I don't see how these changes could affect eth_syncing. Also, I am not sure we should write such a general test here.
In case of a delegated node we don't run les protocol at all and this is why we won't have this data available. The test should ensure that `status-react` can still access this data.
Of course, this test should be commented out and don't run as a regular test suite but it may be worth writing. You've done this check manually but the same could be done with a test.

> I fixed what I found relevant to these changes but this command prints insane amount of alerts, even in internal go files. We should revise its purpose and just select the most important rules.
Yeah, its not easy now, that's why #281

tiabc commented:
By the way, one of the side effects of #272 is `AccountsListRequestHandler` isn't registered during `StatusBackend.registerHandlers` call: https://github.com/status-im/status-go/blob/9f57cf422f46dbd52a11bcc7971204b9d748ec4d/geth/api/backend.go#L208

Due to that our `eth_accounts` call processing won't trigger post processing for accounts and won't append sub-accounts. Could you also fix this under your PR?

adambabik commented:
@tiabc I fixed that test. It was because I though `To` is required, but it can be `nil` in fact. Also, I added more logs so that we can observe how transaction progressed. Also, I can still confirm that that test sometimes fails because of `no suitable peers found` (there is a comment about that added by you a while ago).

influx6 commented on geth/jail/execution_policy.go:
```diff
@@ -3,148 +3,83 @@ package jail
+	case params.SendTransactionMethodName:
+		return ep.executeSendTransaction(req, call)
+	default:
+		return ep.executeOtherTransaction(req, call)
```
You might need to fix conflict merges if #276  gets merged before this PR.

adambabik replied:
Yup, I am prepared for that :)

tiabc commented:
`build/env.sh go test -v ./geth/api/... -testify.m TestJailSendQueuedTransaction` never finishes, checked 3 times.
Can you, please, run all modified tests one-by-one and ensure they work? (at least, sometimes; when we don't face weird errors)
I'd suggest merging `develop` first.

tiabc commented on geth/api/backend_jail_test.go:
```diff
@@ -249,9 +251,13 @@ func (s *BackendTestSuite) TestContractDeployment() {
-				responseValue = contract.transactionHash
+				if (!contract.address) {
+					responseValue = contract.transactionHash;
+				}
```
Could you elaborate on this? How did it work before?

adambabik replied:
This is probably not required anymore. This test was failing for me, but it turned out that it was a bug in the Go code. However, I left it as while debugging I found out how this callback works here https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethcontract.

tiabc commented on geth/node/txqueue_manager_test.go:
```diff
@@ -0,0 +1,218 @@
+		_, err := txQueueManager.CompleteTransaction(tx.ID, TestConfig.Account1.Password)
+		if err != errTxAssumedSent {
+			s.Fail(err.Error())
+		}
```
`testify.Equal`?

adambabik commented:
@tiabc I fixed the test you mentioned and also walked through all affected tests one by one. All of them pass (some are flaky but they were marked as such before).
